library/Notifications/Hook/EventsObjectsInfoHook.php

```diff
@@ -0,0 +1,119 @@
+{
+    /** @var array<string, ValidHtml>  */
+    protected static $objectNameWidgets = [];
+
+    /**
+     * Get the object name widgets for the objects using the object ID tags
+     *
+     * @param array<string, array<string, string>> $objectIdTags
+     *
+     * @return array<string, ValidHtml>
+     */
+    abstract public function getObjectNameWidgets(array $objectIdTags): array;
+
+    /**
+     * Get the source of the objects
+     *
+     * @return string
+     */
+    abstract public function getSource(): string;
+
+    /**
+     * Get the object list item widget for the given object ID tag
+     *
+     * @param array<string, string> $objectIdTag
+     *
+     * @return ValidHtml
+     */
+    abstract public function getHtmlForObject(array $objectIdTag): ValidHtml;
+
+    /**
+     * Register object name widgets for the objects using the object ID tags
+     *
+     * @param array<string, array<string, array<string, string>>> $objectIdTags
+     *
+     * @return void
+     */
+    final public static function registerObjectNameWidgets(array $objectIdTags): void
+    {
+        $objectDisplayNames = [];
+        foreach (Hook::all('Notifications\\EventsObjectsInfo') as $hook) {
+            /** @var self $hook */
+            try {
+                $objectDisplayNames = array_merge(
+                    $objectDisplayNames,
+                    $hook->getObjectNameWidgets($objectIdTags[$hook->getSource()])
+                );
+            } catch (Exception $e) {
+                Logger::error('Failed to load hook %s:', get_class($hook), $e);
+            }
+        }
+
+        self::$objectNameWidgets = $objectDisplayNames;
+    }
+
+    /**
+     * Get the display name for the given object
+     *
+     * @param Objects $obj
+     *
+     * @return ValidHtml
+     */
+    final public static function getObjectNameWidget(Objects $obj): ValidHtml
+    {
+        /** @var Query|Source $source */
+        $source = $obj->source;
+        if ($source instanceof Query) {
+            $source = $obj->source->first();
+        }
+
+        $objectTags = [];
+
+        foreach ($obj->id_tags as $tag => $value) {
+            $objectTags[] = sprintf('%s=%s', $tag, $value);
+        }
+
+        return self::$objectNameWidgets[$source->type][$obj->id] ?? new HtmlString(implode(', ', $objectTags));
+    }
+
+    /**
+     * Get the object list item widget for the object using the object ID tag
+     *
+     * @param string $source
+     * @param array<string, string> $objectIdTag
+     *
+     * @return ?ValidHtml
+     */
+    final public static function getObjectListItemWidget(string $source, array $objectIdTag): ?ValidHtml
+    {
+        $objectListItemWidget = null;
+        foreach (Hook::all('Notifications\\EventsObjectsInfo') as $hook) {
+            /** @var self $hook */
+            try {
+                if ($source === $hook->getSource()) {
+                    $objectListItemWidget = $hook->getHtmlForObject($objectIdTag);
```

Comment:
You have still not returned in the loop. Should the first `$objectListItemWidget` found be overwritten if there is more than one hook?

Reply:
No, every object as a single source so this will not be replaced and that is also the reason for checking whether source's hook matches with object's hook. But I guess it is still better to return here.

Reply:
Yes, because further iterations are useless.